### DISCUSSION TITLE
Add CLAP_PLUGIN_FEATURE_MIDI_EFFECT "midi-effect" to plugin-features.h

### DISCUSSION THREAD

**Discussion by elanhickler** [free-audio/clap#368] [Ideas]
Add CLAP_PLUGIN_FEATURE_MIDI_EFFECT "midi-effect" to plugin-features.h

This is a category for plugins that manipulate and output incoming midi stuff. Could be notes, could be doing something with CC values such as translating to another CC, doing some math like multiply for CC gain or whathaveyou.

Now, I don't know if "midi" is the right word anymore because we have "CLAP_NOTE_DIALECT_CLAP" vs "CLAP_NOTE_DIALECT_MIDI". So, whatever word represents general note, CC, pitchwheel, etc. What word is that?

**Comment by baconpaul** [comment]
We’ve used note effect to map to a midi effect in wrappers and that’s how you make a clap show up as blue in bitwig. I agree it is more than notes

**Reply by abique** [reply]
Maybe note-effect isn't the best terminology. But that's what we have and I think it's OK for it to also represent midi effects.

**Comment by elanhickler** [comment]
I don't know if there's a good one word that represents all midi data without using the word midi.

event-effect
expression-effect
phrase-effect
articulation-effect
score-effect
notation-effect
data-effect
riff-effect

**Reply by elanhickler** [reply]
assuming a plugin can be in multiple categories (ignoring subcategories), then note-effect and cc-effect as two seperate things might work, however `cc-effect` kinda sucks as a word and `continuous-controller-effect` is pretty bad as well.

**Reply by abique** [reply]
This isn't user facing data, it doesn't matter, but I'm not convinced there is a real benefit in adding those.

**Reply by elanhickler** [reply]
rewriting this...

When it comes to categories, you want to have a suggested standard for all common things, otherwise you'll have midi plugins with 10 different names for the same category. Take a look at the original code with my added comments

```cpp
// AUDIO GENERATOR
// Add this feature if your plugin can process note events and then produce audio
#define CLAP_PLUGIN_FEATURE_INSTRUMENT "instrument" 

// AUDIO PROCESSOR
// Add this feature if your plugin is an audio effect
#define CLAP_PLUGIN_FEATURE_AUDIO_EFFECT "audio-effect" // audio proccessor

// AUDIO ANALYZER
// Add this feature if your plugin is an analyzer
#define CLAP_PLUGIN_FEATURE_ANALYZER "analyzer"

// MIDI GENERATOR & PROCESSOR  <-- here we have an inconsistency, these should be seperate OR
// if not seperate, it should be clarified that this category is for everything related to midi/expression
// Add this feature if your plugin is a note effect or a note generator/sequencer // midi generator
#define CLAP_PLUGIN_FEATURE_NOTE_EFFECT "note-effect"

// This category is strange, probably should be a sub-category of pitch-detector with a category of note-generator
// Add this feature if your plugin converts audio to notes
#define CLAP_PLUGIN_FEATURE_NOTE_DETECTOR "note-detector"

// We're missing a midi analyzer category... or CLAP_PLUGIN_FEATURE_ANALYZER should be clarified to include midi
```
Essentially it boils down to `generator or processor or analyzer` and `audio or data`. In the analog world, this is the difference between AC and DC or audio vs control voltage.

By the way, this is the reason no one organizes their plugins via category. If you don't have a good category list, then the category feature will go unused by most people. I've wanted to use categories in the past. CLAP is in its early days, we have a chance to make the category system standardized and useful.

**Comment by elanhickler** [comment]
This is written with the assumption that categories and sub-categories are used like a tagging system so you can have multiple categories and subcategories combined to produce a good definition of a plugin while making it filterable/searchable.

Ok, here's a practical suggestion that is based on the original and taking into account usefulness to the end user. ~~All midi stuff is in one category because I'm assuming we don't really need more than one because I can't think of a use case where a midi plugin only generates instead of generating based on input.~~ nevermind, thought of some.

It seems the main use of categories is to answer the question "does this plugin go in the instrument slot or the effects slot? Does it work on its own or does it need a source to function? Does it go after audio or after midi? Does it generate audio or data or is it just for visuals?" 

Once we answer this question, then we have sub-categories to answer the question "how does it sound, what does it analyze, what does it do?"

```cpp
// CATEGORIES [where does it go in my project?]
instrument // synthesizers, samplers
audio-effect // distortions, delays
analyzer // scopes, graphs, indicators, meters, including midi analyzers

midi-generator // chord suggestors and other interface-only midi plugins
midi-effect // note generators, arpeggiators, chorders, cc processors

utility // notepad, text generator, clocks, file browser, image previewer, interface-only plugins
```

...perhaps move `utility` out of `sub-category` unless a plugin can simply not have a category and only a sub-category

```cpp
// SUB-CATEGORIES (that are missing) [what does it do?]
auto-tuner // pitch quantizers, pitch correction
arpeggiator //  midi note generators
chorder // midi chord generators
multiband // multiband compressor, multiband distortion
fft // spectral paint, spectral filter, denoiser, declipper, dereverb
lo-fi // sample-rate and bit-rate reduction, quantizers, decimators, tape emulation, vinyl emulation, 
convolution // impulse response-based reverb, cabinet, speaker, microphone emulation
```

**Reply by elanhickler** [reply]
hmm, midi-generator sounds like a category for an arpeggiator and devs will certainly make that mistake based on my suggestion. Here's a new proposal. We can keep the note-generator title and simply add midi-effect as suggested in the original post.

```cpp
note-generator // arpeggiators, chord generators
midi-effect // continuous controller manipulation
```
